src/appleseed/foundation/image/oiioimagefilewriter.cpp

```diff
@@ -0,0 +1,515 @@
+        type.arraylen = 8;
+
+        spec.attribute("chromaticities", type, chromaticities);
+    }
+}
+
+void OIIOImageFileWriter::set_generic_image_attributes(const ImageAttributes& image_attributes)
+{
+    OIIO::ImageSpec spec = m_images->m_spec.back();
+
+    for (const_each<ImageAttributes> i = image_attributes; i; ++i)
+    {
+        // Fetch the name and the value of the attribute.
+        const std::string attr_name = i->key();
+        const std::string attr_value = i->value<std::string>();
+
+        if (attr_name == "author")
+            spec.attribute("Artist", attr_value.c_str());
+
+        else if (attr_name == "copyright")
+            spec.attribute("Copyright", attr_value.c_str());
+
+        else if (attr_name == "title")
+            spec.attribute("DocumentName", attr_value.c_str());
+
+        else if (attr_name == "description")
+            spec.attribute("ImageDescription", attr_value.c_str());
+
+        else if (attr_name == "date")
+            spec.attribute("DateTime", attr_value.c_str());
+
+        else if (attr_name == "software")
+            spec.attribute("Software", attr_value.c_str());
+
+        else if (attr_name == "computer")
+            spec.attribute("HostComputer", attr_value.c_str());
+
+        else if (attr_name == "image_name")
+            spec.attribute("oiio:subimagename", attr_value.c_str());
+
+        else if (attr_name == "color_space")
+        {
+            if (attr_value == "linear")
+                spec.attribute("oiio:ColorSpace", "Linear");
+            else
+                spec.attribute("oiio:ColorSpace", attr_value.c_str());
+        }
+
+        else if (attr_name == "compression")
+            spec.attribute("compression", attr_value.c_str());
+
+        else if (attr_name == "compression_quality")
+            spec.attribute("CompressionQuality", from_string<int>(attr_value));
+
+        else if (attr_name == "dpi")
+        {
+            const size_t dpi = from_string<size_t>(attr_value);
+            const double dpm = dpi * (100.0 / 2.54);
+            spec.attribute("XResolution", static_cast<float>(dpm));
+            spec.attribute("YResolution", static_cast<float>(dpm));
+            spec.attribute("ResolutionUnit", "cm");
+        }
+    }
+}
+
+void OIIOImageFileWriter::set_image_attributes(const ImageAttributes& image_attributes)
+{
+    if (get_image_count() == 0)
+        throw ExceptionIOError("No images available!");
+
+    // Retrieve filename extension
+    const boost::filesystem::path filepath(m_filename);
+    const std::string extension = lower_case(filepath.extension().string());
+
+    // General image attributes
+    set_generic_image_attributes(image_attributes);
+
+    // Set image attributes depending of its extension
+    if (extension == ".exr")
+        set_exr_image_attributes(image_attributes);
+}
+
+void OIIOImageFileWriter::write_tiles(const size_t image_index)
+{
+    assert(image_index < m_images->m_canvas.size());
```

Comment:
Redundant assertion, it's already checked just below.

Reply:
done